File: attachment_upload.py
"""Bounded binary attachment uploads over the existing Bridge WebSocket."""
from __future__ import annotations

import hashlib
import json
import os
import re
import subprocess
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Awaitable, Callable


BINARY_MAGIC = b"CBV1"
UPLOAD_ID_BYTES = 34  # ``u_`` + 32 hex characters
MAX_VIDEO_BYTES = 512 * 1024 * 1024
VIDEO_EXTENSIONS = {
    ".mp4", ".mov", ".m4v", ".webm", ".mkv", ".avi", ".3gp", ".3gpp",
}

SendJson = Callable[[dict[str, Any]], Awaitable[Any]]


def _safe_component(value: str, fallback: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "_", value).strip("._")
    return (cleaned[:120] or fallback)


def _video_metadata(path: Path) -> dict[str, Any]:
    """Return best-effort ffprobe metadata without making uploads depend on it."""
    try:
        proc = subprocess.run(
            [
                "ffprobe", "-v", "error", "-show_entries",
                "format=duration:stream=width,height,codec_name",
                "-select_streams", "v:0", "-of", "json", str(path),
            ],
            capture_output=True,
            text=True,
            timeout=15,
            check=False,
        )
        if proc.returncode != 0:
            return {}
        raw = json.loads(proc.stdout or "{}")
        stream = (raw.get("streams") or [{}])[0]
        duration = float((raw.get("format") or {}).get("duration") or 0)
        result: dict[str, Any] = {}
        if duration > 0:
            result["duration_ms"] = round(duration * 1000)
        for key in ("width", "height", "codec_name"):
            if stream.get(key) is not None:
                result[key] = stream[key]
        return result
    except (OSError, ValueError, json.JSONDecodeError, subprocess.SubprocessError):
        return {}


@dataclass
class _Upload:
    upload_id: str
    request_id: str
    session_id: str
    device_id: str
    name: str
    media_type: str
    expected_size: int
    part_path: Path
    final_path: Path
    handle: Any
    digest: Any
    received: int = 0


class ConnectionUploadManager:
    """Owns in-progress uploads for one authenticated connection."""

    def __init__(self, data_dir: str, sessions: dict[str, Any], device_id: str):
        base = Path(data_dir or os.path.expanduser("~/.claude-bridge-runtime"))
        self.upload_root = (base / "uploads").resolve()
        self.sessions = sessions
        self.device_id = _safe_component(device_id, "unknown-device")
        self.active: dict[str, _Upload] = {}

    async def init(self, msg: dict[str, Any], send_json: SendJson) -> None:
        request_id = str(msg.get("upload_request_id") or "")
        session_id = str(msg.get("session_id") or "")
        name = str(msg.get("name") or "recording.mp4")
        media_type = str(msg.get("media_type") or "video/mp4")
        try:
            size = int(msg.get("size_bytes"))
        except (TypeError, ValueError):
            size = -1

        suffix = Path(name).suffix.lower()
        error = ""
        if not request_id:
            error = "Missing upload_request_id"
        elif session_id not in self.sessions:
            error = "Unknown session"
        elif size <= 0:
            error = "Video is empty"
        elif size > MAX_VIDEO_BYTES:
            error = "Video exceeds the 512 MB limit"
        elif suffix not in VIDEO_EXTENSIONS or not media_type.startswith("video/"):
            error = "Unsupported video format"
        if error:
            await send_json({
                "type": "attachment_upload_error",
                "upload_request_id": request_id,
                "message": error,
            })
            return

        upload_id = f"u_{uuid.uuid4().hex}"
        safe_session = _safe_component(session_id, "session")
        safe_name = _safe_component(Path(name).stem, "recording") + suffix
        target_dir = self.upload_root / safe_session / upload_id
        target_dir.mkdir(parents=True, exist_ok=False)
        part_path = target_dir / f".{safe_name}.part"
        final_path = target_dir / safe_name
        handle = part_path.open("xb")
        self.active[upload_id] = _Upload(
            upload_id=upload_id,
            request_id=request_id,
            session_id=session_id,
            device_id=self.device_id,
            name=safe_name,
            media_type=media_type,
            expected_size=size,
            part_path=part_path,
            final_path=final_path,
            handle=handle,
            digest=hashlib.sha256(),
        )
        await send_json({
            "type": "attachment_upload_ready",
            "upload_request_id": request_id,
            "upload_id": upload_id,
            "chunk_size": 256 * 1024,
        })

    async def binary(self, frame: bytes, send_json: SendJson) -> None:
        header_len = len(BINARY_MAGIC) + UPLOAD_ID_BYTES
        if len(frame) <= header_len or not frame.startswith(BINARY_MAGIC):
            return
        try:
            upload_id = frame[len(BINARY_MAGIC):header_len].decode("ascii")
        except UnicodeDecodeError:
            return
        upload = self.active.get(upload_id)
        if upload is None:
            return
        chunk = frame[header_len:]
        if upload.received + len(chunk) > upload.expected_size:
            await self._fail(upload, send_json, "Received more bytes than declared")
            return
        upload.handle.write(chunk)
        upload.digest.update(chunk)
        upload.received += len(chunk)

    async def finish(self, msg: dict[str, Any], send_json: SendJson) -> None:
        upload_id = str(msg.get("upload_id") or "")
        upload = self.active.get(upload_id)
        if upload is None:
            await send_json({
                "type": "attachment_upload_error",
                "upload_id": upload_id,
                "message": "Unknown or expired upload",
            })
            return
        if upload.received != upload.expected_size:
            await self._fail(
                upload,
                send_json,
                f"Incomplete upload ({upload.received}/{upload.expected_size} bytes)",
            )
            return
        upload.handle.flush()
        os.fsync(upload.handle.fileno())
        upload.handle.close()
        upload.part_path.replace(upload.final_path)
        metadata = _video_metadata(upload.final_path)
        manifest = {
            "version": 1,
            "upload_id": upload.upload_id,
            "session_id": upload.session_id,
            "device_id": upload.device_id,
            "name": upload.name,
            "media_type": upload.media_type,
            "size_bytes": upload.expected_size,
            "sha256": upload.digest.hexdigest(),
            "path": str(upload.final_path),
            **metadata,
        }
        manifest_path = upload.final_path.parent / "manifest.json"
        manifest_path.write_text(
            json.dumps(manifest, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )
        self.active.pop(upload_id, None)
        await send_json({
            "type": "attachment_upload_complete",
            "upload_request_id": upload.request_id,
            **manifest,
            "remote_path": str(upload.final_path),
        })

    async def cancel(self, msg: dict[str, Any], send_json: SendJson | None = None) -> None:
        upload_id = str(msg.get("upload_id") or "")
        upload = self.active.get(upload_id)
        if upload is not None:
            self._discard(upload)

    def close(self) -> None:
        for upload in list(self.active.values()):
            self._discard(upload)

    async def _fail(self, upload: _Upload, send_json: SendJson, message: str) -> None:
        self._discard(upload)
        await send_json({
            "type": "attachment_upload_error",
            "upload_request_id": upload.request_id,
            "upload_id": upload.upload_id,
            "message": message,
        })

    def _discard(self, upload: _Upload) -> None:
        self.active.pop(upload.upload_id, None)
        try:
            upload.handle.close()
        except Exception:
            pass
        try:
            upload.part_path.unlink(missing_ok=True)
            upload.part_path.parent.rmdir()
        except OSError:
            pass


def resolve_uploaded_videos(
    files: Any,
    *,
    session_id: str,
    data_dir: str,
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """Split inline files from validated uploaded video references."""
    if not isinstance(files, list):
        return [], []
    upload_root = (Path(data_dir or os.path.expanduser("~/.claude-bridge-runtime")) / "uploads").resolve()
    inline: list[dict[str, Any]] = []
    videos: list[dict[str, Any]] = []
    for item in files:
        if not isinstance(item, dict) or not item.get("remote_path"):
            inline.append(item)
            continue
        candidate = Path(str(item["remote_path"])).resolve()
        try:
            candidate.relative_to(upload_root)
        except ValueError as exc:
            raise ValueError("Uploaded video path is outside the attachment store") from exc
        manifest_path = candidate.parent / "manifest.json"
        try:
            manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            raise ValueError("Uploaded video manifest is missing or invalid") from exc
        if manifest.get("session_id") != session_id or Path(str(manifest.get("path"))).resolve() != candidate:
            raise ValueError("Uploaded video does not belong to this session")
        if not candidate.is_file() or candidate.stat().st_size != int(manifest.get("size_bytes") or -1):
            raise ValueError("Uploaded video is missing or incomplete")
        videos.append(manifest)
    return inline, videos


def append_video_context(content: str, videos: list[dict[str, Any]]) -> str:
    if not videos:
        return content
    lines = [
        "",
        "",
        "[Bridge attached video files — inspect the original files at these absolute paths]",
    ]
    for video in videos:
        details = [
            f"name={video.get('name')}",
            f"path={video.get('path')}",
            f"media_type={video.get('media_type')}",
            f"size_bytes={video.get('size_bytes')}",
            f"sha256={video.get('sha256')}",
        ]
        if video.get("duration_ms"):
            details.append(f"duration_ms={video['duration_ms']}")
        if video.get("width") and video.get("height"):
            details.append(f"dimensions={video['width']}x{video['height']}")
        lines.append("- " + "; ".join(details))
    lines.append(
        "Use local video tools (for example ffprobe/ffmpeg frame extraction) to inspect them; "
        "do not claim to have reviewed the video unless you actually opened or sampled it."
    )
    return content + "\n".join(lines)

File: internal/core/attachment_upload.go
package core

import (
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"hash"
	"io"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"everything-go/internal/backend"
)

const (
	attachmentMagicV1     = "CBV1"
	attachmentMagicV2     = "CBV2"
	attachmentIDBytes     = 34
	attachmentOffsetBytes = 8
	maxVideoUploadBytes   = int64(512 * 1024 * 1024)
	attachmentChunkBytes  = 512 * 1024
	staleUploadMaxAge     = 24 * time.Hour
)

var safeAttachmentComponent = regexp.MustCompile(`[^A-Za-z0-9._-]+`)

type activeAttachmentUpload struct {
	id, requestID, sessionID, deviceID string
	name, mediaType                    string
	expected, received                 int64
	partPath, finalPath, statePath     string
	file                               *os.File
	digest                             hash.Hash
}

type attachmentUploads struct {
	root    string
	client  *Client
	active  map[string]*activeAttachmentUpload
	cleaned bool
}

type uploadedVideoManifest struct {
	Version    int    `json:"version"`
	UploadID   string `json:"upload_id"`
	SessionID  string `json:"session_id"`
	DeviceID   string `json:"device_id"`
	Name       string `json:"name"`
	MediaType  string `json:"media_type"`
	SizeBytes  int64  `json:"size_bytes"`
	SHA256     string `json:"sha256"`
	Path       string `json:"path"`
	DurationMS int64  `json:"duration_ms,omitempty"`
	Width      int    `json:"width,omitempty"`
	Height     int    `json:"height,omitempty"`
}

type resumableUploadState struct {
	Version      int    `json:"version"`
	UploadID     string `json:"upload_id"`
	RequestID    string `json:"upload_request_id"`
	SessionID    string `json:"session_id"`
	DeviceID     string `json:"device_id"`
	Name         string `json:"name"`
	MediaType    string `json:"media_type"`
	ExpectedSize int64  `json:"expected_size"`
	ReceivedSize int64  `json:"received_size"`
	PartPath     string `json:"part_path"`
	FinalPath    string `json:"final_path"`
}

func newAttachmentUploads(c *Client, dataDir string) *attachmentUploads {
	return &attachmentUploads{
		root: filepath.Join(dataDir, "uploads"), client: c,
		active: make(map[string]*activeAttachmentUpload),
	}
}

func attachmentSafe(value, fallback string) string {
	value = strings.Trim(safeAttachmentComponent.ReplaceAllString(value, "_"), "._")
	if value == "" {
		return fallback
	}
	if len(value) > 120 {
		return value[:120]
	}
	return value
}

func isVideoExtension(ext string) bool {
	switch strings.ToLower(ext) {
	case ".mp4", ".mov", ".m4v", ".webm", ".mkv", ".avi", ".3gp", ".3gpp":
		return true
	default:
		return false
	}
}

func resumableAttachmentID(deviceID, requestID string) string {
	sum := sha256.Sum256([]byte(deviceID + "\x00" + requestID))
	return "u_" + hex.EncodeToString(sum[:16])
}

func (u *attachmentUploads) init(sessionID, requestID, name, mediaType string, size int64) {
	if !u.cleaned {
		u.cleanupStale()
		u.cleaned = true
	}
	if requestID == "" {
		u.sendError(requestID, "", "Missing upload_request_id")
		return
	}
	if _, ok := u.client.hub.registry.Get(sessionID); !ok {
		u.sendError(requestID, "", "Unknown session")
		return
	}
	ext := strings.ToLower(filepath.Ext(name))
	if size <= 0 {
		u.sendError(requestID, "", "Video is empty")
		return
	}
	if size > maxVideoUploadBytes {
		u.sendError(requestID, "", "Video exceeds the 512 MB limit")
		return
	}
	if !strings.HasPrefix(mediaType, "video/") || !isVideoExtension(ext) {
		u.sendError(requestID, "", "Unsupported video format")
		return
	}
	id := resumableAttachmentID(u.client.deviceID, requestID)
	if existing := u.active[id]; existing != nil {
		u.client.enqueueEvent(map[string]any{
			"type": "attachment_upload_ready", "upload_request_id": requestID,
			"upload_id": id, "chunk_size": attachmentChunkBytes,
			"protocol_version": 2, "received_bytes": existing.received,
		})
		return
	}
	safeName := attachmentSafe(strings.TrimSuffix(filepath.Base(name), filepath.Ext(name)), "recording") + ext
	dir := filepath.Join(u.root, attachmentSafe(sessionID, "session"), id)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		u.sendError(requestID, id, "Cannot create attachment store")
		return
	}
	partPath := filepath.Join(dir, "."+safeName+".part")
	finalPath := filepath.Join(dir, safeName)
	statePath := filepath.Join(dir, "upload-state.json")

	if complete, ok := u.completedUpload(requestID, sessionID, id, finalPath); ok {
		u.client.enqueueEvent(complete)
		return
	}

	received, digest, err := u.loadResumableState(
		statePath, partPath, id, requestID, sessionID, u.client.deviceID,
		safeName, mediaType, size,
	)
	if err != nil {
		u.sendError(requestID, id, err.Error())
		return
	}
	f, err := os.OpenFile(partPath, os.O_WRONLY|os.O_CREATE, 0o600)
	if err != nil {
		u.sendError(requestID, id, "Cannot create upload file")
		return
	}
	if _, err := f.Seek(received, io.SeekStart); err != nil {
		_ = f.Close()
		u.sendError(requestID, id, "Cannot resume upload file")
		return
	}
	u.active[id] = &activeAttachmentUpload{
		id: id, requestID: requestID, sessionID: sessionID, deviceID: u.client.deviceID,
		name: safeName, mediaType: mediaType, expected: size, received: received,
		partPath: partPath, finalPath: finalPath, statePath: statePath,
		file: f, digest: digest,
	}
	if err := u.persistState(u.active[id]); err != nil {
		u.discard(u.active[id])
		u.sendError(requestID, id, "Cannot save upload state")
		return
	}
	u.client.uploadActive.Store(true)
	u.client.enqueueEvent(map[string]any{
		"type": "attachment_upload_ready", "upload_request_id": requestID,
		"upload_id": id, "chunk_size": attachmentChunkBytes,
		"protocol_version": 2, "received_bytes": received,
	})
}

func (u *attachmentUploads) cleanupStale() {
	cutoff := time.Now().Add(-staleUploadMaxAge)
	var staleDirs []string
	_ = filepath.WalkDir(u.root, func(path string, entry fs.DirEntry, err error) error {
		if err != nil || entry.IsDir() || entry.Name() != "upload-state.json" {
			return nil
		}
		info, statErr := entry.Info()
		if statErr == nil && info.ModTime().Before(cutoff) {
			staleDirs = append(staleDirs, filepath.Dir(path))
		}
		return nil
	})
	for _, dir := range staleDirs {
		if _, err := os.Stat(filepath.Join(dir, "manifest.json")); errors.Is(err, os.ErrNotExist) {
			_ = os.RemoveAll(dir)
		}
	}
}

func (u *attachmentUploads) writeFrame(data []byte) bool {
	if len(data) < len(attachmentMagicV1) {
		return false
	}
	magic := string(data[:len(attachmentMagicV1)])
	if magic != attachmentMagicV1 && magic != attachmentMagicV2 {
		return false
	}
	header := len(attachmentMagicV1) + attachmentIDBytes
	if magic == attachmentMagicV2 {
		header += attachmentOffsetBytes
	}
	if len(data) <= header {
		return true
	}
	id := string(data[len(attachmentMagicV1) : len(attachmentMagicV1)+attachmentIDBytes])
	up := u.active[id]
	if up == nil {
		return true
	}
	if magic == attachmentMagicV2 {
		offset := int64(binary.BigEndian.Uint64(data[header-attachmentOffsetBytes : header]))
		if offset != up.received {
			u.sendAck(up)
			return true
		}
	}
	chunk := data[header:]
	if up.received+int64(len(chunk)) > up.expected {
		u.fail(up, "Received more bytes than declared")
		return true
	}
	n, err := up.file.Write(chunk)
	if err != nil || n != len(chunk) {
		u.fail(up, "Failed to write upload")
		return true
	}
	_, _ = up.digest.Write(chunk)
	up.received += int64(len(chunk))
	if err := up.file.Sync(); err != nil {
		u.fail(up, "Failed to sync upload")
		return true
	}
	if err := u.persistState(up); err != nil {
		u.fail(up, "Failed to save upload progress")
		return true
	}
	if magic == attachmentMagicV2 {
		u.sendAck(up)
	}
	return true
}

func (u *attachmentUploads) finish(id string) {
	up := u.active[id]
	if up == nil {
		u.sendError("", id, "Unknown or expired upload")
		return
	}
	if up.received != up.expected {
		u.fail(up, fmt.Sprintf("Incomplete upload (%d/%d bytes)", up.received, up.expected))
		return
	}
	if err := up.file.Sync(); err != nil {
		u.fail(up, "Failed to sync upload")
		return
	}
	if err := up.file.Close(); err != nil {
		u.fail(up, "Failed to close upload")
		return
	}
	up.file = nil
	if err := os.Rename(up.partPath, up.finalPath); err != nil {
		u.fail(up, "Failed to finalize upload")
		return
	}
	manifest := uploadedVideoManifest{
		Version: 1, UploadID: up.id, SessionID: up.sessionID, DeviceID: up.deviceID,
		Name: up.name, MediaType: up.mediaType, SizeBytes: up.expected,
		SHA256: hex.EncodeToString(up.digest.Sum(nil)), Path: up.finalPath,
	}
	manifest.DurationMS, manifest.Width, manifest.Height = probeVideo(up.finalPath)
	raw, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil || os.WriteFile(filepath.Join(filepath.Dir(up.finalPath), "manifest.json"), raw, 0o600) != nil {
		_ = os.Remove(up.finalPath)
		u.fail(up, "Failed to save upload manifest")
		return
	}
	_ = os.Remove(up.statePath)
	delete(u.active, id)
	u.client.uploadActive.Store(len(u.active) > 0)
	u.client.enqueueEvent(map[string]any{
		"type": "attachment_upload_complete", "upload_request_id": up.requestID,
		"upload_id": up.id, "name": up.name, "media_type": up.mediaType,
		"size_bytes": up.expected, "sha256": manifest.SHA256,
		"remote_path": up.finalPath, "path": up.finalPath,
		"duration_ms": manifest.DurationMS, "width": manifest.Width, "height": manifest.Height,
	})
}

func probeVideo(path string) (int64, int, int) {
	out, err := exec.Command(
		"ffprobe", "-v", "error", "-show_entries", "format=duration:stream=width,height",
		"-select_streams", "v:0", "-of", "json", path,
	).Output()
	if err != nil {
		return 0, 0, 0
	}
	var result struct {
		Streams []struct{ Width, Height int } `json:"streams"`
		Format  struct{ Duration string }     `json:"format"`
	}
	if json.Unmarshal(out, &result) != nil {
		return 0, 0, 0
	}
	seconds, _ := strconv.ParseFloat(result.Format.Duration, 64)
	if len(result.Streams) == 0 {
		return int64(seconds * 1000), 0, 0
	}
	return int64(seconds * 1000), result.Streams[0].Width, result.Streams[0].Height
}

func (u *attachmentUploads) cancel(id string) {
	if up := u.active[id]; up != nil {
		u.discard(up)
	}
	u.client.uploadActive.Store(len(u.active) > 0)
}

func (u *attachmentUploads) close() {
	for _, up := range u.active {
		if up.file != nil {
			_ = up.file.Sync()
			_ = up.file.Close()
			up.file = nil
		}
		_ = u.persistState(up)
		delete(u.active, up.id)
	}
	u.client.uploadActive.Store(false)
}

func (u *attachmentUploads) fail(up *activeAttachmentUpload, message string) {
	requestID, id := up.requestID, up.id
	u.discard(up)
	u.client.uploadActive.Store(len(u.active) > 0)
	u.sendError(requestID, id, message)
}

func (u *attachmentUploads) discard(up *activeAttachmentUpload) {
	delete(u.active, up.id)
	if up.file != nil {
		_ = up.file.Close()
	}
	_ = os.Remove(up.partPath)
	_ = os.Remove(up.statePath)
	_ = os.Remove(filepath.Dir(up.partPath))
}

func (u *attachmentUploads) sendAck(up *activeAttachmentUpload) {
	u.client.enqueueEvent(map[string]any{
		"type": "attachment_upload_ack", "upload_request_id": up.requestID,
		"upload_id": up.id, "received_bytes": up.received,
	})
}

func (u *attachmentUploads) persistState(up *activeAttachmentUpload) error {
	state := resumableUploadState{
		Version: 2, UploadID: up.id, RequestID: up.requestID,
		SessionID: up.sessionID, DeviceID: up.deviceID,
		Name: up.name, MediaType: up.mediaType,
		ExpectedSize: up.expected, ReceivedSize: up.received,
		PartPath: up.partPath, FinalPath: up.finalPath,
	}
	raw, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	tmp := up.statePath + ".tmp"
	if err := os.WriteFile(tmp, raw, 0o600); err != nil {
		return err
	}
	return os.Rename(tmp, up.statePath)
}

func (u *attachmentUploads) loadResumableState(
	statePath, partPath, id, requestID, sessionID, deviceID, name, mediaType string,
	expected int64,
) (int64, hash.Hash, error) {
	digest := sha256.New()
	raw, err := os.ReadFile(statePath)
	if errors.Is(err, os.ErrNotExist) {
		if info, statErr := os.Stat(partPath); statErr == nil && info.Size() > 0 {
			return 0, nil, errors.New("Upload state is missing")
		}
		return 0, digest, nil
	}
	if err != nil {
		return 0, nil, errors.New("Cannot read upload state")
	}
	var state resumableUploadState
	if json.Unmarshal(raw, &state) != nil ||
		state.Version != 2 ||
		state.UploadID != id ||
		state.RequestID != requestID ||
		state.SessionID != sessionID ||
		state.DeviceID != deviceID ||
		state.Name != name ||
		state.MediaType != mediaType ||
		state.ExpectedSize != expected ||
		state.PartPath != partPath {
		return 0, nil, errors.New("Upload resume metadata does not match")
	}
	f, err := os.Open(partPath)
	if err != nil {
		return 0, nil, errors.New("Upload partial file is missing")
	}
	defer f.Close()
	n, err := io.Copy(digest, f)
	if err != nil || n != state.ReceivedSize || n > expected {
		return 0, nil, errors.New("Upload partial file is inconsistent")
	}
	return n, digest, nil
}

func (u *attachmentUploads) completedUpload(
	requestID, sessionID, id, finalPath string,
) (map[string]any, bool) {
	raw, err := os.ReadFile(filepath.Join(filepath.Dir(finalPath), "manifest.json"))
	if err != nil {
		return nil, false
	}
	var manifest uploadedVideoManifest
	if json.Unmarshal(raw, &manifest) != nil ||
		manifest.UploadID != id ||
		manifest.SessionID != sessionID ||
		manifest.Path != finalPath {
		return nil, false
	}
	info, err := os.Stat(finalPath)
	if err != nil || info.Size() != manifest.SizeBytes {
		return nil, false
	}
	return map[string]any{
		"type": "attachment_upload_complete", "upload_request_id": requestID,
		"upload_id": manifest.UploadID, "name": manifest.Name,
		"media_type": manifest.MediaType, "size_bytes": manifest.SizeBytes,
		"sha256": manifest.SHA256, "remote_path": manifest.Path, "path": manifest.Path,
		"duration_ms": manifest.DurationMS, "width": manifest.Width, "height": manifest.Height,
	}, true
}

func (u *attachmentUploads) sendError(requestID, id, message string) {
	u.client.enqueueEvent(map[string]any{
		"type": "attachment_upload_error", "upload_request_id": requestID,
		"upload_id": id, "message": message,
	})
}

func (h *Hub) resolveUploadedVideos(sessionID, content string, files []backend.FileAttachment) (string, []backend.FileAttachment, error) {
	inline := make([]backend.FileAttachment, 0, len(files))
	var videos []uploadedVideoManifest
	root, err := filepath.Abs(filepath.Join(h.cfg.DataDir, "uploads"))
	if err != nil {
		return content, nil, err
	}
	for _, file := range files {
		if file.RemotePath == "" {
			inline = append(inline, file)
			continue
		}
		candidate, err := filepath.Abs(file.RemotePath)
		if err != nil {
			return content, nil, err
		}
		rel, err := filepath.Rel(root, candidate)
		if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(os.PathSeparator)) {
			return content, nil, errors.New("uploaded video path is outside the attachment store")
		}
		raw, err := os.ReadFile(filepath.Join(filepath.Dir(candidate), "manifest.json"))
		if err != nil {
			return content, nil, errors.New("uploaded video manifest is missing")
		}
		var manifest uploadedVideoManifest
		if json.Unmarshal(raw, &manifest) != nil || manifest.SessionID != sessionID || manifest.Path != candidate {
			return content, nil, errors.New("uploaded video does not belong to this session")
		}
		info, err := os.Stat(candidate)
		if err != nil || info.Size() != manifest.SizeBytes {
			return content, nil, errors.New("uploaded video is missing or incomplete")
		}
		videos = append(videos, manifest)
	}
	if len(videos) == 0 {
		return content, inline, nil
	}
	var b strings.Builder
	b.WriteString(content)
	b.WriteString("\n\n[Bridge attached video files — inspect the original files at these absolute paths]\n")
	for _, video := range videos {
		fmt.Fprintf(&b, "- name=%s; path=%s; media_type=%s; size_bytes=%d; sha256=%s",
			video.Name, video.Path, video.MediaType, video.SizeBytes, video.SHA256)
		if video.DurationMS > 0 {
			fmt.Fprintf(&b, "; duration_ms=%d", video.DurationMS)
		}
		if video.Width > 0 && video.Height > 0 {
			fmt.Fprintf(&b, "; dimensions=%dx%d", video.Width, video.Height)
		}
		b.WriteByte('\n')
	}
	b.WriteString("Use local video tools (for example ffprobe/ffmpeg frame extraction) to inspect them; do not claim to have reviewed the video unless you actually opened or sampled it.")
	return b.String(), inline, nil
}
